File: packages/doge-sdk-typescript/src/platform.ts
import type { DogeClient } from './client.js'
import type {
  Capability,
  CapabilitySnapshot,
  CaseAssetLink,
  CaseDecision,
  CaseProgressStep,
  CaseRunLink,
  CaseReview,
  HomeQueue,
  Project,
  ResearchCase,
  SlotInstallResponse,
  SlotStatusRow,
  TemplatePreflightResult,
  WorkflowExecution,
  WorkflowTemplate,
  Workspace,
} from './platform-types.js'

interface ListProjectsOptions {
  workspaceId?: string
  limit?: number
}

interface ListResearchCasesOptions {
  projectId?: string
  limit?: number
}

interface CreateWorkflowTemplateOptions {
  description?: string
  currentVersion?: string
  inputSchema?: Record<string, unknown>
  runInstructions?: string
  toolPolicy?: Record<string, unknown>
  evidencePolicy?: Record<string, unknown>
  outputContract?: Record<string, unknown>
  metadata?: Record<string, unknown>
  requiredCapabilities?: string[]
  evalPolicy?: string[]
  approvalPolicy?: Record<string, unknown>
  uiSchema?: Record<string, unknown>
}

interface CreateResearchCaseRunFromTemplateOptions {
  question?: string
  modelPolicy?: Record<string, unknown>
  inputs?: Record<string, unknown>
  workflow?: string
  sessionId?: string
  market?: string
  language?: string
  documentIds?: string[]
  portfolioId?: string
  linkType?: string
}

interface CaseExecutionOptions {
  question?: string
  workflow?: string
  sessionId?: string
  market?: string
  language?: string
  documentIds?: string[]
  portfolioId?: string
  assetLinkIds?: string[]
  modelPolicy?: Record<string, unknown>
  inputs?: Record<string, unknown>
  skipPreflight?: boolean
}

export class PlatformResource {
  private readonly root: DogeClient

  constructor(root: DogeClient) {
    this.root = root
  }

  async listWorkspaces(limit = 100): Promise<Workspace[]> {
    const payload = await this.root.request<{ workspaces: Workspace[] }>(
      'GET',
      `/v1/workspaces?limit=${limit}`,
    )
    return payload.workspaces
  }

  createWorkspace(name: string, description = ''): Promise<Workspace> {
    return this.root.request<Workspace>('POST', '/v1/workspaces', { name, description })
  }

  getWorkspace(workspaceId: string): Promise<Workspace> {
    return this.root.request<Workspace>('GET', `/v1/workspaces/${workspaceId}`)
  }

  async listProjects(options: ListProjectsOptions = {}): Promise<Project[]> {
    const payload = await this.root.request<{ projects: Project[] }>(
      'GET',
      queryPath('/v1/projects', { workspace_id: options.workspaceId, limit: options.limit ?? 100 }),
    )
    return payload.projects
  }

  createProject(
    workspaceId: string,
    name: string,
    options: { description?: string, defaultMarket?: string } = {},
  ): Promise<Project> {
    return this.root.request<Project>('POST', '/v1/projects', {
      workspace_id: workspaceId,
      name,
      description: options.description ?? '',
      default_market: options.defaultMarket,
    })
  }

  getProject(projectId: string): Promise<Project> {
    return this.root.request<Project>('GET', `/v1/projects/${projectId}`)
  }

  async listResearchCases(options: ListResearchCasesOptions = {}): Promise<ResearchCase[]> {
    const payload = await this.root.request<{ research_cases: ResearchCase[] }>(
      'GET',
      queryPath('/v1/research-cases', { project_id: options.projectId, limit: options.limit ?? 100 }),
    )
    return payload.research_cases
  }

  createResearchCase(projectId: string, title: string, thesis = ''): Promise<ResearchCase> {
    return this.root.request<ResearchCase>('POST', '/v1/research-cases', { project_id: projectId, title, thesis })
  }

  getResearchCase(caseId: string): Promise<ResearchCase> {
    return this.root.request<ResearchCase>('GET', `/v1/research-cases/${caseId}`)
  }

  async getCaseProgress(caseId: string): Promise<CaseProgressStep[]> {
    const payload = await this.root.request<{ steps: CaseProgressStep[] }>(
      'GET',
      `/v1/research-cases/${caseId}/progress`,
    )
    return payload.steps
  }

  homeQueue(limit = 20): Promise<HomeQueue> {
    return this.root.request<HomeQueue>('GET', `/v1/home-queue?limit=${limit}`)
  }

  linkResearchCaseRun(caseId: string, runId: string, linkType = 'primary'): Promise<CaseRunLink> {
    return this.root.request<CaseRunLink>('POST', `/v1/research-cases/${caseId}/runs`, {
      run_id: runId,
      link_type: linkType,
    })
  }

  createResearchCaseRunFromTemplate(
    caseId: string,
    templateId: string,
    options: CreateResearchCaseRunFromTemplateOptions = {},
  ): Promise<CaseRunLink> {
    return this.root.request<CaseRunLink>('POST', `/v1/research-cases/${caseId}/runs`, {
      template_id: templateId,
      question: options.question,
      model_policy: options.modelPolicy ?? {},
      inputs: options.inputs ?? {},
      workflow: options.workflow,
      session_id: options.sessionId,
      market: options.market ?? 'us',
      language: options.language ?? 'en',
      document_ids: options.documentIds ?? [],
      portfolio_id: options.portfolioId,
      link_type: options.linkType ?? 'primary',
    })
  }

  async listCaseAssets(caseId: string): Promise<CaseAssetLink[]> {
    const payload = await this.root.request<{ assets: CaseAssetLink[] }>(
      'GET',
      `/v1/research-cases/${caseId}/assets`,
    )
    return payload.assets
  }

  addCaseAsset(
    caseId: string,
    assetType: string,
    assetId: string,
    options: {
      assetName?: string
      role?: string
      version?: string
      metadata?: Record<string, unknown>
    } = {},
  ): Promise<CaseAssetLink> {
    return this.root.request<CaseAssetLink>('POST', `/v1/research-cases/${caseId}/assets`, {
      asset_type: assetType,
      asset_id: assetId,
      asset_name: options.assetName ?? '',
      role: options.role ?? 'source',
      version: options.version,
      metadata: options.metadata ?? {},
    })
  }

  async listCaseDecisions(caseId: string): Promise<CaseDecision[]> {
    const payload = await this.root.request<{ decisions: CaseDecision[] }>(
      'GET',
      `/v1/research-cases/${caseId}/decisions`,
    )
    return payload.decisions
  }

  recordCaseDecision(
    caseId: string,
    decisionType: string,
    options: {
      rationale?: string
      sourceRunIds?: string[]
      sourceExecutionIds?: string[]
    } = {},
  ): Promise<CaseDecision> {
    return this.root.request<CaseDecision>('POST', `/v1/research-cases/${caseId}/decisions`, {
      decision_type: decisionType,
      rationale: options.rationale ?? '',
      source_run_ids: options.sourceRunIds ?? [],
      source_execution_ids: options.sourceExecutionIds ?? [],
    })
  }

  preflightCaseExecution(
    caseId: string,
    templateId: string,
    options: CaseExecutionOptions = {},
  ): Promise<TemplatePreflightResult> {
    return this.root.request<TemplatePreflightResult>(
      'POST',
      `/v1/research-cases/${caseId}/executions/preflight`,
      caseExecutionPayload(templateId, options),
    )
  }

  executeCaseTemplate(
    caseId: string,
    templateId: string,
    options: CaseExecutionOptions = {},
  ): Promise<WorkflowExecution> {
    return this.root.request<WorkflowExecution>(
      'POST',
      `/v1/research-cases/${caseId}/executions`,
      caseExecutionPayload(templateId, options),
    )
  }

  async listCaseExecutions(caseId: string, limit = 100): Promise<WorkflowExecution[]> {
    const payload = await this.root.request<{ executions: WorkflowExecution[] }>(
      'GET',
      `/v1/research-cases/${caseId}/executions?limit=${limit}`,
    )
    return payload.executions
  }

  getCaseReview(caseId: string): Promise<CaseReview> {
    return this.root.request<CaseReview>('GET', `/v1/research-cases/${caseId}/review`)
  }

  async listWorkflowTemplates(limit = 100): Promise<WorkflowTemplate[]> {
    const payload = await this.root.request<{ workflow_templates: WorkflowTemplate[] }>(
      'GET',
      `/v1/workflow-templates?limit=${limit}`,
    )
    return payload.workflow_templates
  }

  createWorkflowTemplate(
    slug: string,
    name: string,
    options: CreateWorkflowTemplateOptions = {},
  ): Promise<WorkflowTemplate> {
    return this.root.request<WorkflowTemplate>('POST', '/v1/workflow-templates', {
      slug,
      name,
      description: options.description ?? '',
      current_version: options.currentVersion ?? '1',
      input_schema: options.inputSchema ?? {},
      run_instructions: options.runInstructions ?? '',
      tool_policy: options.toolPolicy ?? {},
      evidence_policy: options.evidencePolicy ?? {},
      output_contract: options.outputContract ?? {},
      metadata: options.metadata ?? {},
      required_capabilities: options.requiredCapabilities,
      eval_policy: options.evalPolicy,
      approval_policy: options.approvalPolicy,
      ui_schema: options.uiSchema,
    })
  }

  getWorkflowTemplate(templateId: string): Promise<WorkflowTemplate> {
    return this.root.request<WorkflowTemplate>('GET', `/v1/workflow-templates/${templateId}`)
  }

  async listSlots(): Promise<SlotStatusRow[]> {
    const payload = await this.root.request<{ slots: SlotStatusRow[] }>('GET', '/v1/slots')
    return payload.slots
  }

  getSlot(slotId: string): Promise<SlotStatusRow> {
    return this.root.request<SlotStatusRow>('GET', `/v1/slots/${slotId}`)
  }

  installSlot(source: string): Promise<SlotInstallResponse> {
    return this.root.request<SlotInstallResponse>('POST', '/v1/slots/install', { source })
  }
}

export class CapabilitiesResource {
  private readonly root: DogeClient

  constructor(root: DogeClient) {
    this.root = root
  }

  get(): Promise<CapabilitySnapshot> {
    return this.root.request<CapabilitySnapshot>('GET', '/v1/capabilities')
  }

  async list(): Promise<Capability[]> {
    const payload = await this.root.request<{ capabilities: Capability[] }>('GET', '/v1/capabilities')
    return payload.capabilities
  }
}

function queryPath(path: string, params: Record<string, string | number | undefined>): string {
  const query = new URLSearchParams()
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined) query.set(key, String(value))
  }
  const suffix = query.toString()
  return suffix ? `${path}?${suffix}` : path
}

function caseExecutionPayload(templateId: string, options: CaseExecutionOptions): Record<string, unknown> {
  return {
    template_id: templateId,
    question: options.question,
    workflow: options.workflow,
    session_id: options.sessionId,
    market: options.market ?? 'us',
    language: options.language ?? 'en',
    document_ids: options.documentIds ?? [],
    portfolio_id: options.portfolioId,
    asset_link_ids: options.assetLinkIds ?? [],
    model_policy: options.modelPolicy ?? {},
    inputs: options.inputs ?? {},
    skip_preflight: options.skipPreflight ?? false,
  }
}

File: packages/doge-sdk-python/doge_sdk/platform.py
"""Platform and capability resources for the Python SDK."""

from __future__ import annotations

from typing import Any


class PlatformResource:
    def __init__(self, root: Any) -> None:
        self._root = root

    def list_workspaces(self, limit: int = 100) -> list[dict[str, Any]]:
        return self._root._request("GET", "/v1/workspaces", params={"limit": limit})["workspaces"]

    def create_workspace(self, name: str, description: str = "") -> dict[str, Any]:
        return self._root._request("POST", "/v1/workspaces", json={"name": name, "description": description})

    def get_workspace(self, workspace_id: str) -> dict[str, Any]:
        return self._root._request("GET", f"/v1/workspaces/{workspace_id}")

    def list_projects(self, workspace_id: str | None = None, limit: int = 100) -> list[dict[str, Any]]:
        params: dict[str, Any] = {"limit": limit}
        if workspace_id is not None:
            params["workspace_id"] = workspace_id
        return self._root._request("GET", "/v1/projects", params=params)["projects"]

    def create_project(
        self,
        workspace_id: str,
        name: str,
        description: str = "",
        default_market: str | None = None,
    ) -> dict[str, Any]:
        return self._root._request(
            "POST",
            "/v1/projects",
            json={
                "workspace_id": workspace_id,
                "name": name,
                "description": description,
                "default_market": default_market,
            },
        )

    def get_project(self, project_id: str) -> dict[str, Any]:
        return self._root._request("GET", f"/v1/projects/{project_id}")

    def list_research_cases(self, project_id: str | None = None, limit: int = 100) -> list[dict[str, Any]]:
        params: dict[str, Any] = {"limit": limit}
        if project_id is not None:
            params["project_id"] = project_id
        return self._root._request("GET", "/v1/research-cases", params=params)["research_cases"]

    def create_research_case(self, project_id: str, title: str, thesis: str = "") -> dict[str, Any]:
        return self._root._request(
            "POST",
            "/v1/research-cases",
            json={"project_id": project_id, "title": title, "thesis": thesis},
        )

    def get_research_case(self, case_id: str) -> dict[str, Any]:
        return self._root._request("GET", f"/v1/research-cases/{case_id}")

    def home_queue(self, limit: int = 20) -> dict[str, Any]:
        return self._root._request("GET", "/v1/home-queue", params={"limit": limit})

    def link_research_case_run(self, case_id: str, run_id: str, link_type: str = "primary") -> dict[str, Any]:
        return self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/runs",
            json={"run_id": run_id, "link_type": link_type},
        )

    def create_research_case_run_from_template(
        self,
        case_id: str,
        template_id: str,
        *,
        question: str | None = None,
        model_policy: dict[str, Any] | None = None,
        inputs: dict[str, Any] | None = None,
        workflow: str | None = None,
        session_id: str | None = None,
        market: str = "us",
        language: str = "en",
        document_ids: list[str] | None = None,
        portfolio_id: str | None = None,
        link_type: str = "primary",
    ) -> dict[str, Any]:
        return self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/runs",
            json={
                "template_id": template_id,
                "question": question,
                "model_policy": model_policy or {},
                "inputs": inputs or {},
                "workflow": workflow,
                "session_id": session_id,
                "market": market,
                "language": language,
                "document_ids": document_ids or [],
                "portfolio_id": portfolio_id,
                "link_type": link_type,
            },
        )

    def list_case_assets(self, case_id: str) -> list[dict[str, Any]]:
        return self._root._request("GET", f"/v1/research-cases/{case_id}/assets")["assets"]

    def add_case_asset(
        self,
        case_id: str,
        asset_type: str,
        asset_id: str,
        *,
        asset_name: str = "",
        role: str = "source",
        version: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/assets",
            json={
                "asset_type": asset_type,
                "asset_id": asset_id,
                "asset_name": asset_name,
                "role": role,
                "version": version,
                "metadata": metadata or {},
            },
        )

    def list_case_decisions(self, case_id: str) -> list[dict[str, Any]]:
        return self._root._request("GET", f"/v1/research-cases/{case_id}/decisions")["decisions"]

    def record_case_decision(
        self,
        case_id: str,
        decision_type: str,
        *,
        rationale: str = "",
        source_run_ids: list[str] | None = None,
        source_execution_ids: list[str] | None = None,
    ) -> dict[str, Any]:
        return self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/decisions",
            json={
                "decision_type": decision_type,
                "rationale": rationale,
                "source_run_ids": source_run_ids or [],
                "source_execution_ids": source_execution_ids or [],
            },
        )

    def preflight_case_execution(self, case_id: str, template_id: str, **kwargs: Any) -> dict[str, Any]:
        return self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/executions/preflight",
            json=_case_execution_payload(template_id, kwargs),
        )

    def execute_case_template(self, case_id: str, template_id: str, **kwargs: Any) -> dict[str, Any]:
        return self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/executions",
            json=_case_execution_payload(template_id, kwargs),
        )

    def list_case_executions(self, case_id: str, limit: int = 100) -> list[dict[str, Any]]:
        return self._root._request(
            "GET",
            f"/v1/research-cases/{case_id}/executions",
            params={"limit": limit},
        )["executions"]

    def get_case_review(self, case_id: str) -> dict[str, Any]:
        return self._root._request("GET", f"/v1/research-cases/{case_id}/review")

    def list_workflow_templates(self, limit: int = 100) -> list[dict[str, Any]]:
        return self._root._request("GET", "/v1/workflow-templates", params={"limit": limit})["workflow_templates"]

    def create_workflow_template(
        self,
        slug: str,
        name: str,
        description: str = "",
        current_version: str = "1",
        input_schema: dict[str, Any] | None = None,
        run_instructions: str = "",
        tool_policy: dict[str, Any] | None = None,
        evidence_policy: dict[str, Any] | None = None,
        output_contract: dict[str, Any] | None = None,
        metadata: dict[str, Any] | None = None,
        required_capabilities: list[str] | None = None,
        eval_policy: list[str] | None = None,
        approval_policy: dict[str, Any] | None = None,
        ui_schema: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return self._root._request(
            "POST",
            "/v1/workflow-templates",
            json={
                "slug": slug,
                "name": name,
                "description": description,
                "current_version": current_version,
                "input_schema": input_schema or {},
                "run_instructions": run_instructions,
                "tool_policy": tool_policy or {},
                "evidence_policy": evidence_policy or {},
                "output_contract": output_contract or {},
                "metadata": metadata or {},
                "required_capabilities": required_capabilities,
                "eval_policy": eval_policy,
                "approval_policy": approval_policy,
                "ui_schema": ui_schema,
            },
        )

    def get_workflow_template(self, template_id: str) -> dict[str, Any]:
        return self._root._request("GET", f"/v1/workflow-templates/{template_id}")


class CapabilitiesResource:
    def __init__(self, root: Any) -> None:
        self._root = root

    def get(self) -> dict[str, Any]:
        return self._root._request("GET", "/v1/capabilities")

    def list(self) -> list[dict[str, Any]]:
        return self.get()["capabilities"]


class AsyncPlatformResource:
    def __init__(self, root: Any) -> None:
        self._root = root

    async def list_workspaces(self, limit: int = 100) -> list[dict[str, Any]]:
        return (await self._root._request("GET", "/v1/workspaces", params={"limit": limit}))["workspaces"]

    async def create_workspace(self, name: str, description: str = "") -> dict[str, Any]:
        return await self._root._request("POST", "/v1/workspaces", json={"name": name, "description": description})

    async def get_workspace(self, workspace_id: str) -> dict[str, Any]:
        return await self._root._request("GET", f"/v1/workspaces/{workspace_id}")

    async def list_projects(self, workspace_id: str | None = None, limit: int = 100) -> list[dict[str, Any]]:
        params: dict[str, Any] = {"limit": limit}
        if workspace_id is not None:
            params["workspace_id"] = workspace_id
        return (await self._root._request("GET", "/v1/projects", params=params))["projects"]

    async def create_project(
        self,
        workspace_id: str,
        name: str,
        description: str = "",
        default_market: str | None = None,
    ) -> dict[str, Any]:
        return await self._root._request(
            "POST",
            "/v1/projects",
            json={
                "workspace_id": workspace_id,
                "name": name,
                "description": description,
                "default_market": default_market,
            },
        )

    async def get_project(self, project_id: str) -> dict[str, Any]:
        return await self._root._request("GET", f"/v1/projects/{project_id}")

    async def list_research_cases(self, project_id: str | None = None, limit: int = 100) -> list[dict[str, Any]]:
        params: dict[str, Any] = {"limit": limit}
        if project_id is not None:
            params["project_id"] = project_id
        return (await self._root._request("GET", "/v1/research-cases", params=params))["research_cases"]

    async def create_research_case(self, project_id: str, title: str, thesis: str = "") -> dict[str, Any]:
        return await self._root._request(
            "POST",
            "/v1/research-cases",
            json={"project_id": project_id, "title": title, "thesis": thesis},
        )

    async def get_research_case(self, case_id: str) -> dict[str, Any]:
        return await self._root._request("GET", f"/v1/research-cases/{case_id}")

    async def home_queue(self, limit: int = 20) -> dict[str, Any]:
        return await self._root._request("GET", "/v1/home-queue", params={"limit": limit})

    async def link_research_case_run(self, case_id: str, run_id: str, link_type: str = "primary") -> dict[str, Any]:
        return await self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/runs",
            json={"run_id": run_id, "link_type": link_type},
        )

    async def create_research_case_run_from_template(
        self,
        case_id: str,
        template_id: str,
        *,
        question: str | None = None,
        model_policy: dict[str, Any] | None = None,
        inputs: dict[str, Any] | None = None,
        workflow: str | None = None,
        session_id: str | None = None,
        market: str = "us",
        language: str = "en",
        document_ids: list[str] | None = None,
        portfolio_id: str | None = None,
        link_type: str = "primary",
    ) -> dict[str, Any]:
        return await self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/runs",
            json={
                "template_id": template_id,
                "question": question,
                "model_policy": model_policy or {},
                "inputs": inputs or {},
                "workflow": workflow,
                "session_id": session_id,
                "market": market,
                "language": language,
                "document_ids": document_ids or [],
                "portfolio_id": portfolio_id,
                "link_type": link_type,
            },
        )

    async def list_case_assets(self, case_id: str) -> list[dict[str, Any]]:
        return (await self._root._request("GET", f"/v1/research-cases/{case_id}/assets"))["assets"]

    async def add_case_asset(
        self,
        case_id: str,
        asset_type: str,
        asset_id: str,
        *,
        asset_name: str = "",
        role: str = "source",
        version: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return await self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/assets",
            json={
                "asset_type": asset_type,
                "asset_id": asset_id,
                "asset_name": asset_name,
                "role": role,
                "version": version,
                "metadata": metadata or {},
            },
        )

    async def list_case_decisions(self, case_id: str) -> list[dict[str, Any]]:
        return (await self._root._request("GET", f"/v1/research-cases/{case_id}/decisions"))["decisions"]

    async def record_case_decision(
        self,
        case_id: str,
        decision_type: str,
        *,
        rationale: str = "",
        source_run_ids: list[str] | None = None,
        source_execution_ids: list[str] | None = None,
    ) -> dict[str, Any]:
        return await self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/decisions",
            json={
                "decision_type": decision_type,
                "rationale": rationale,
                "source_run_ids": source_run_ids or [],
                "source_execution_ids": source_execution_ids or [],
            },
        )

    async def preflight_case_execution(self, case_id: str, template_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/executions/preflight",
            json=_case_execution_payload(template_id, kwargs),
        )

    async def execute_case_template(self, case_id: str, template_id: str, **kwargs: Any) -> dict[str, Any]:
        return await self._root._request(
            "POST",
            f"/v1/research-cases/{case_id}/executions",
            json=_case_execution_payload(template_id, kwargs),
        )

    async def list_case_executions(self, case_id: str, limit: int = 100) -> list[dict[str, Any]]:
        return (await self._root._request(
            "GET",
            f"/v1/research-cases/{case_id}/executions",
            params={"limit": limit},
        ))["executions"]

    async def get_case_review(self, case_id: str) -> dict[str, Any]:
        return await self._root._request("GET", f"/v1/research-cases/{case_id}/review")

    async def list_workflow_templates(self, limit: int = 100) -> list[dict[str, Any]]:
        return (
            await self._root._request("GET", "/v1/workflow-templates", params={"limit": limit})
        )["workflow_templates"]

    async def create_workflow_template(
        self,
        slug: str,
        name: str,
        description: str = "",
        current_version: str = "1",
        input_schema: dict[str, Any] | None = None,
        run_instructions: str = "",
        tool_policy: dict[str, Any] | None = None,
        evidence_policy: dict[str, Any] | None = None,
        output_contract: dict[str, Any] | None = None,
        metadata: dict[str, Any] | None = None,
        required_capabilities: list[str] | None = None,
        eval_policy: list[str] | None = None,
        approval_policy: dict[str, Any] | None = None,
        ui_schema: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return await self._root._request(
            "POST",
            "/v1/workflow-templates",
            json={
                "slug": slug,
                "name": name,
                "description": description,
                "current_version": current_version,
                "input_schema": input_schema or {},
                "run_instructions": run_instructions,
                "tool_policy": tool_policy or {},
                "evidence_policy": evidence_policy or {},
                "output_contract": output_contract or {},
                "metadata": metadata or {},
                "required_capabilities": required_capabilities,
                "eval_policy": eval_policy,
                "approval_policy": approval_policy,
                "ui_schema": ui_schema,
            },
        )

    async def get_workflow_template(self, template_id: str) -> dict[str, Any]:
        return await self._root._request("GET", f"/v1/workflow-templates/{template_id}")


class AsyncCapabilitiesResource:
    def __init__(self, root: Any) -> None:
        self._root = root

    async def get(self) -> dict[str, Any]:
        return await self._root._request("GET", "/v1/capabilities")

    async def list(self) -> list[dict[str, Any]]:
        return (await self.get())["capabilities"]


def _case_execution_payload(template_id: str, options: dict[str, Any]) -> dict[str, Any]:
    return {
        "template_id": template_id,
        "question": options.get("question"),
        "workflow": options.get("workflow"),
        "session_id": options.get("session_id"),
        "market": options.get("market", "us"),
        "language": options.get("language", "en"),
        "document_ids": options.get("document_ids") or [],
        "portfolio_id": options.get("portfolio_id"),
        "asset_link_ids": options.get("asset_link_ids") or [],
        "model_policy": options.get("model_policy") or {},
        "inputs": options.get("inputs") or {},
        "skip_preflight": options.get("skip_preflight", False),
    }
